### DISCUSSION TITLE
[hardhat-zksolc]<Error: write EPIPE>

### DISCUSSION THREAD

**Discussion by Confucian-e** [zkSync-Community-Hub/zksync-developers#355] [Compilers] [answered]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

solc v0.8.19

### zksolc/zkvyper version?

zksolc v1.3.23

### Explain the issue

I'm trying to compile using zksolc but it shows error

```bash
pnpm hardhat compile --network zkSyncTestnet
```

```bash
Compiling contracts for zkSync Era with zksolc v1.3.23 and solc v0.8.19
Compiling 14 Solidity files
Error: write EPIPE
    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16)
```

**Comment by dutterbutter** [comment]
@Confucian-e Can you please provide more details on the issue? A repo to reproduce would be ideal.

**Reply by Confucian-e** [reply]
Well my `hardhat.config.ts` file is:

```typescript
import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";
import "@nomicfoundation/hardhat-toolbox";
import { HardhatUserConfig } from "hardhat/config";

const config: HardhatUserConfig = {
  networks: {
    hardhat: {
      zksync: false,
    },
    zkSyncTestnet: {
      url: "https://sepolia.era.zksync.dev",
      ethNetwork: "sepolia", // or a Sepolia RPC endpoint from Infura/Alchemy/Chainstack etc.
      zksync: true,
    },
  },
  zksolc: {
    version: "latest",
    settings: {},
  },
  solidity: {
    version: "0.8.19",
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      }
    }
  },
};

export default config;
```

**Comment by Confucian-e** [comment]
Well I just tried your cli, but it shows same error
I followed tutorial here https://docs.zksync.io/build/tooling/hardhat/getting-started.html#project-setup

First I run 
```bash
pnpm dlx zksync-cli create demo --template hardhat_solidity
```

Then run
```bash
pnpm compile
```
Then it shows
```bash

> zksync-hardhat-template@ compile /Users/confucian/dev/zksync/demo
> hardhat compile

Compiling contracts for zkSync Era with zksolc v1.3.23 and solc v0.8.17
Compiling 48 Solidity files
Error: write EPIPE
    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16)
 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by boyuanx** [comment]
Getting the same issue with v0.8.24

**Comment by ljada37** [comment]
The error write EPIPE typically indicates an issue with writing to a closed pipe. This could be caused by various factors such as incorrect file permissions, insufficient disk space, or a bug in the tooling. To resolve this issue, you can try the following steps:

Ensure that you have sufficient disk space available on your system.
Check the file permissions for the directories and files involved in the compilation process.
Make sure that all dependencies are installed correctly and up to date.
If the issue persists, consider restarting your development environment or reinstalling the tooling to rule out any temporary glitches.
If none of the above steps resolve the issue, you may need to investigate further or seek assistance from the tooling's community or support channels.

**Comment by kakarukjoseph** [comment]
The error "write EPIPE" typically indicates a broken pipe error, which can occur when there's an issue with writing to a closed pipe or socket. This error message suggests that there might be an issue with the compilation process, possibly due to a problem with the input or output streams.

To address this, you could try the following steps:
1. Ensure that your contracts are correctly written and do not contain any syntax errors.
2. Check if there are any issues with your hardhat configuration or network settings.
3. Verify that all necessary dependencies are correctly installed and up to date.
4. Try running the compilation command with additional verbosity or debugging options to gather more information about the cause of the error.

If the issue persists, you may need to further investigate the specific circumstances under which the error occurs or seek assistance from the hardhat or zkSync community for troubleshooting.

**Comment by boyuanx** [comment, marked as answer]
zksolc 1.4.0 fixed this issue.

**Comment by tumb2ibonsa** [comment]
The error "write EPIPE" typically occurs when a process tries to write to a pipe that has been closed. This could be due to various reasons, such as a problem with the file system, insufficient permissions, or a bug in the compiler or build tool. To resolve the issue, you may need to investigate further to identify the root cause.

**Comment by dyksterhouset** [comment]
Compiler: zksolc
Solidity Version: v0.8.19
zksolc Version: v1.3.23
Issue: Error message "write EPIPE" encountered while compiling contracts.

**Comment by anggabarber** [comment]
It seems like there is an issue with compiling Solidity files using zksolc v1.3.23 and solc v0.8.19. The error message "write EPIPE" is being displayed, which indicates a problem with writing to a pipe. Without further information or access to the code repository, it is difficult to diagnose the exact issue.

**Comment by davosloper** [comment]
I am using 1.5.1 (also tried 1.4.0) zksolc and I am still getting EPIPE and sometimes EOF error when I compile. I have updated all packages and node version. Any idea why ?

**Reply by davosloper** [reply]
![image](https://github.com/user-attachments/assets/73fc879e-b2aa-4bc5-9bb0-e7caac295443)
![image](https://github.com/user-attachments/assets/c085a881-34f1-4463-b25c-fb0cc8bae810)

**Reply by bxpana** [reply]
Do you have a repo link you can share? Any luck when trying with 1.5.0? or 1.4.1?

**Reply by davosloper** [reply]
I don't have repository link for this. I just followed the migration part in the docs and tried with 1.3.22, 1.4.0, 1.5.1. Nothing works for me.

**Reply by bxpana** [reply]
Which docs are you using and without a minimal reproducible code it's hard to replicate to see what's going on. Is this on a local network or Sepolia?

**Reply by davosloper** [reply]
I followed here: https://docs.zksync.io/build/tooling/hardhat/migrating-to-zksync

Also tried on MacOS 12 (Virtual Machine) and got the following error:
![Screenshot 2024-08-15 015200](https://github.com/user-attachments/assets/5aab4688-36b7-4f75-868b-a780d29a3d06)

**Reply by bxpana** [reply]
@davosloper Let's open up another discussion around this so we can get as much information about your project as we can. Like I said before it's hard to replicate without any code or a repo
